src/backend/sgx/enclave.rs

```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Enclave object
+
+use std::fs::{File, OpenOptions};
+use std::io::Result;
+
+pub const SGX_ENCLAVE_DEVICE: &str = "/dev/sgx_enclave";
+
+/// Wraps an enclave file descriptor.
+pub struct Enclave(File);
+
+impl From<Enclave> for File {
+    fn from(enclave: Enclave) -> File {
+        enclave.0
+    }
+}
```

Comment:
Maybe implement `DerefMut` so you can transparently use `Enclave` as if it was a `File`.

Reply:
So I tried this:
```rust
impl Deref for Enclave {
    type Target = File;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for Enclave {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl AsRawFd for Enclave {
    fn as_raw_fd(&self) -> RawFd {
        self.0.as_raw_fd()
    }
}

impl Enclave {
    pub fn new() -> Result<Enclave> {
        match OpenOptions::new()
            .read(true)
            .write(true)
            .open(SGX_ENCLAVE_DEVICE)
        {
            Ok(file) => Ok(Enclave(file)),
            Err(err) => Err(err),
        }
    }
}
```
Then for `builder.rs`:
```rust
diff --git a/src/backend/sgx/builder.rs b/src/backend/sgx/builder.rs
index c198cc7..204c613 100644
--- a/src/backend/sgx/builder.rs
+++ b/src/backend/sgx/builder.rs
@@ -54,8 +54,7 @@ impl TryFrom<super::config::Config> for Builder {
         // Create a new enclave instance
         let mut file = Enclave::new()
             .context(format!("Failed to open '{}'", SGX_ENCLAVE_DEVICE))
-            .unwrap()
-            .into();
+            .unwrap();
 
         // Create the enclave.
         let secs = config
@@ -73,7 +72,7 @@ impl TryFrom<super::config::Config> for Builder {
             perm: Vec::new(),
             tcsp: Vec::new(),
             cnfg: config,
-            file,
+            file: *file,
         })
     }
 }
```
This results in:
```
$ cargo build
   Compiling enarx v0.3.0 (/home/jarkko/Projects/enarx)
error[E0507]: cannot move out of dereference of `Enclave`
  --> src/backend/sgx/builder.rs:75:19
   |
75 |             file: *file,
   |                   ^^^^^ move occurs because value has type `File`, which does not implement the `Copy` trait

For more information about this error, try `rustc --explain E0507`.
error: could not compile `enarx` due to previous error
```

Reply:
ah, ok, sorry.